Feature/cwa

nbeliy opened the pull request:
Extremely slow

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        if ext[1] != ".cwa":
+            return False
+
+        f = open(filename, "rb")
```
@nbeliy : use `with open(filename, 'rb') as f:` instead. If the opening crashes (or the code crashes without reaching your `close`statement), the file will be closed anyway.

nbeliy replied:
with statement will close the file automatically at the end of block, and I want to keep the stream open

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        if not isinstance(filename, str):
+            raise TypeError("filename must be a string")
+        ext = os.path.splitext(filename)
+        if ext[1] != ".cwa":
```
@nbeliy : `ext[-1]` should be more robust to wild names containing dots. Please add a print statement indicating that the file does not end with the expect extension.

nbeliy replied:
os.path.splitext(path)

    Split the pathname path into a pair (root, ext) such that root + ext == path, and ext is empty or begins with a period and contains at most one period. Leading periods on the basename are ignored; splitext('.cshrc') returns ('.cshrc', '').

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        self.__startTime = datetime.min
+        self.__duration = timedelta()
+        self.__period = 0
+        self.__Accel = pandas.DataFrame(columns=["x","y", "z","norm"])
```
@nbeliy : do not instanciate empty dataframe. Instead, initialize a numpy array with the correct size, fill it and finally, create a dataframe out of it. Appending data to existing dataframe is very slow (if not the slowest method)

See https://stackoverflow.com/questions/10715965/add-one-row-to-pandas-dataframe?rq=1

nbeliy replied:
WIP

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        return False
+
+    def __init__(self, filename, ):
+        if not isinstance(filename, str):
```
@nbeliy : why don't you use your IsValidFile method? Besides, you open and close the file twice (one for checking and one for reading). Can't you do it all at once?

nbeliy replied:
IsValidMethod is for facilitate the determination of file format -- just in master code, you can faslty check if given file is cwa file. It is not intended (but it can) be used in __init__

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        self.__Others = None
+
+        self.__blocks = 0
+        self.__stream.seek(0,2)
```
@nbeliy : add a bit of comments when you do your magic. Here, I had to check the doc to understand you searched for the end of the file.

nbeliy replied:
It's a standard way to determine file size, I seen it as self-explanotary

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        self.__stream.seek(0)
+        if self.__blocks%512 != 0:
+            raise ValueError("Data file corrupted")
+        self.__blocks = self.__blocks//512
```
@nbeliy : do not use the same variable to point to the file position and the number of blocks. The pointer position does not need in that case to be a class attribute. And rename __blocks with __nblocks for the sake of code readability.

nbeliy replied:
__blocks is not a pointer, is an estimation of total number of blocks

ghammad replied:
Here, `self.__blocks = self.__stream.tell() - 1024`, you do store the position in `__blocks` (https://docs.python.org/3/library/io.html#io.IOBase.tell).

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+            if res is None:
+                break
+            if res[4] is not None:
+                self.__Accel = self.__Accel.append(res[4], ignore_index=False)
```
@nbeliy : see comment above about appending row to DF

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        self.printHeader()
+        
+        t = datetime.now()
+        for blk in range(0, self.__blocks):
```
@nbeliy : do I understand correctly that you read 1000 blocks at once and thus reset your counter `rem` every 1000. If so, could you make this parameter '1000' configurable when reading the file? Maybe reading more or less blocks will lead to performance improvement but we need to be able to play with that parameter when reading the file.

nbeliy replied:
It was a test if it will improve the performance

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+        self.__Accel = pandas.DataFrame(columns=["x","y", "z","norm"])
+        self.__Gyro = pandas.DataFrame(columns=["x","y", "z","norm"])
+        self.__Mag = pandas.DataFrame(columns=["x","y", "z","norm"])
+        self.__Ascale = 0
```
@nbeliy : please harmonize names.
Ex:
- `self.__accel`
- `self.__accel_scale`

ghammad commented on pyActigraphy/io/tools.py:
```diff
@@ -0,0 +1,193 @@
+            if precised, data readed at given offset, without
+            advancing internal one
+        """
+        if not isinstance(type, str):
```
@nbeliy : for the sake of code performance, do you think it is necessary to perform these checks every time you call the `unpack` function (which is a lot of times!)? I agree about type checking (although it is maybe very python-oriented) but only when instantiating a class object for example.

nbeliy replied:
I doubt it influence the performance significally. to test

ghammad commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+
+        # @ 2  +2   Packet length (1020 bytes, 
+        # with header (4) = 1024 bytes total)
+        length = dr.unpack("uint16")
```
@nbeliy : in your doc header of the unpack function, you write that providing the offset reads the data without moving the internal pointer. How comes then you manage to unpack the packet length at position @2 without moving the pointer, since the precedent unpacking does use the offset.

More generally, I think that (unless it harms the reading performance) using the offset explicitly in the unpack function makes the code more readable and in-line with the in-line doc of the CWA header structure (which gives the offset and the length of each attribute).

nbeliy replied:
There no strict pointer -- unpack reads data in memory at given index. I simulate the pointer by offset variable, and I decide to move it or not.

Using offset explicitely is error prone: you must track exactly where to read. And in case of packed values an error of one bit not always easely detectable.
That's why I hide offset, as data are readed sequentually. If you need skip some bits, you can always use >>

ghammad replied:
Please re-read the first paragraph of my comment and your doc for your `unpack` function.

For the second paragraph, I meant "using the offset" in the original unpack function (https://docs.python.org/3/library/struct.html#struct.Struct.unpack_from), not yours.

nbeliy commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+import pandas
+from numpy.linalg import norm
+
+class CWA(object):
```
If I need to import only 1 thing from package, I usually do from

ghammad replied:
The comment (as well as the others) refers to the last line:
Why `class CWA(object)` instead of `class CWA()` ?

nbeliy replied:
misreaded. Just to be able to test and debug easyer, without loading all pyActigraphy package

nbeliy commented on pyActigraphy/io/cwa/cwa_class.py:
```diff
@@ -1,237 +1,513 @@
+              }
+
+    @classmethod
+    def IsValidFile(cls, filename):
```
readability is subjective, but feel free to rename

nbeliy commented:
I think it is pointless to discuss the PEP8, variables naming and arguments, as the logic of code is not really ready yet.

ghammad commented:
The PEP8 has nothing to do with the logic of the code. It is easier to write a code that follows that standard from the start, instead of rewriting everything once you are happy with the logic.

nbeliy commented:
## Bottleneck

Is around line 417; calculating the norm (manual is the fastest way) and filling the array. 


## To resolve
Acceleration around 9 g is unrealistic. But even if I manually extract the values from cwa/accsamp.cwa, I retrieve the same numbers. Can you try it manually from the first block?

Accelorimeter:
                                x         y         z       norm
2013-07-13 10:00:07.010 -7.328125  0.125000 -6.703125   9.932217
2013-07-13 10:00:07.010  0.125000 -6.703125 -7.968750  10.413860
2013-07-13 10:00:07.010 -6.703125 -7.968750 -7.890625  13.065022
2013-07-13 10:00:07.010 -7.968750 -7.890625  0.843750  11.246104
2013-07-13 10:00:07.010 -7.890625  0.843750 -7.953125  11.235038

nbeliy commented:
Now the retrieved values are correct!

Remains fixing timestamps, constructor parameters and memory optimisation

nbeliy commented:
I think is a good point to merge.
What is working:

- interface with base class
- correct decoding of values for acti and auxiliary data
- improved performance (take now around 4 min on my machine) 
- implemented timestamps correction
- implemented retrieval of raw (unscaled) data

To do:

- sometimes the timestamps are not ordered (force timestamp correction?)
- improve documentation and code style
- performance (mostly memory) improvement
